xgcm/grid.py

```diff
@@ -861,7 +861,26 @@ def transform(
                 "`transform` can only be used on axes that are non-periodic. Pass `periodic=False` to `xgcm.Grid`."
             )
 
-        def _parse_target(target, target_dim, target_data_dim, target_data):
+        def _target_data_name_handling(target_data):
+            """Handle target_data input without a name"""
+            if target_data.name is None:
+                warnings.warn(
+                    "The input dataarray `target_data` has no name. Will default to `_UNNAMED_`."
+                )
+                da.name = "_UNNAMED_"
+
+        def _check_target_alignment(da, target_da):
+            # check alignment for all other axes, to avoid broadcasting enourmous arrays
+            # to do this on the axis level we will simply check if target_data has some coordinates
+            # that are not found in da (excluding all coordinates that belong to the current axis)
+            da_other_dims = set(da.dims) - set(self.coords.values())
+            target_da_other_dims = set(target_da.dims) - set(self.coords.values())
+            if not target_da_other_dims.issubset(da_other_dims):
+                raise ValueError(
+                    f"Found additional dimensions [{target_da_other_dims-da_other_dims}] in `target_data` not found in `da`. This could mean that the target array is not on the same position along other axes. Use grid.interp() to align the arrays."
```

Comment:
I ran in to this yesterday. `xr.broadcast` or `DataArray.broadcast_like(target_data)` is a solution for some cases.

Also, lets add some newlines in this string :)

Reply:
We actually want to avoid broadcasting on purpose. @rabernat ran into an example where the values were not properly aligned and things just blew up.

Reply:
Sorry I didn't express myself clearly enough. We should mention `broadcast` in the error message as a possible solution. 

Also broadcasting and alignment are two different things. We can check alignment with `xr.align(..., join="exact")` to avoid Ryan's problem. Automatically broadcasting should be fine (apply_ufunc should just do this IIUC) and it'll be obvious from the dimensions on the returned object.

Reply:
Ah ok that makes sense.

Assuming some hypothetical 'da' (x, y, z) and 'target_data'(xc,yc,z), strict
aligning beforehand would just give an empty array, no?

I think that neither an empty array or a resulting (x,xc,y,yc,...) is desirable. Thus I'd prefer to only allow arrays that have identical sims (except along the axis we are transforming), and give an informative warning. 

We could check only the dims which are associated with another axis as a compromise?

@rabernat do you have thoughts on this?

Reply:
Oh I see. Ya you are correct.

> Assuming some hypothetical 'da' (x, y, z) and 'target_data'(xc,yc,z), strict aligning beforehand would just give an empty array, no?

`join="exact"` will raise an error if the z coordinates are not exactly equal on both. But I see now why automatic broadcasting would not work.

My use case was `da` (z) and `target_data` (x, y, z) so I needed to broadcast to get past the check. I wonder if we could augement the if condition to also check `da.dims.issubset(target_data.dims)`, in which case broadcasting is fine.